### PR TITLE
Add max9611 High Side Current Shunt ADC

### PR DESCRIPTION
**Pull request by mckaymatthew**
# What does this implement/fix? 

This adds support for the MAX9611/9612 High Side Current Shunt Amplifier. This is unique in that it supports up to +60VDC common mode. The IC also supports sampling the VDC rail directly as well as die temperature. 
The photo I took is marginal, I do not have a good photography setup. I am not sure how to generate an alternative text/svg for index page, if the image is not acceptable. 

This is my first contribution to ESPHome, I have attempted to follow Contribution Guidelines, apologies if I have missed something. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1612

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
i2c:
  sda: 18
  scl: 19
  scan: true
sensor:
  - platform: max9611
    shunt_resistance: 0.2 ohm
    gain: '1X'
    voltage:
      name: Test Voltage
    current:
      name: Test Current
    power:
      name: Test Watts
    temperature:
      name: Test Temp
    update_interval: 1s

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/max9611/max9611.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    current_sensor_->publish_state(amps);
+  }
+  if (watt_sensor_ != nullptr) {
+    watt_sensor_->publish_state(watts);
+  }
+  if (temperature_sensor_ != nullptr) {
+    temperature_sensor_->publish_state(temp);
+  }
+
+  ESP_LOGI(TAG, "V: %f, A: %f, W: %f, Deg C: %f", voltage, amps, watts, temp);
```
Comment (jesserockz) [inline esphome/components/max9611/max9611.cpp]:
```suggestion
  ESP_LOGD(TAG, "V: %f, A: %f, W: %f, Deg C: %f", voltage, amps, watts, temp);
```

**Comment by mckaymatthew** [comment]
Revert to draft, during formatting something broke. Will resubmit when resolved.

**Comment by mckaymatthew** [comment]
This has been running successfully for a while, resubmitting for PR.

**Review by jesserockz — APPROVED** [review]
Some spacing between functions, namespaces, const blocks would be nice, but otherwise LGTM.

**Pull request merged**